File: src/main/java/com/oycm/algorithm/a/basic/Solution_8.scala
package com.oycm.algorithm.a.basic

object Solution_8 {

  /**
   *
   * 1423.可获得的最大点数 1574
   * https://leetcode.cn/problems/maximum-points-you-can-obtain-from-cards/description/
   * 每次只能从 头或尾 拿一张卡牌，求拿出恰好 k 张牌的点数和最大值
   * @param cardPoints 整数数组，表示排成一行卡牌的点数
   * @param k 整数，恰好 k 张牌
   * @return
   */
  def maxScore(cardPoints: Array[Int], k: Int): Int = {
    var ans = 0
    var temp = 0
    var left = 0
    var right = cardPoints.length
    var flag = false

    // 可拆分两个 for k 循环，写法更简单
    // 从左边先去 k 张牌求和，去掉头最右边的牌再依次取尾部的牌
    while (left < right && left >= 0) {

      if (!flag) {
        // 左边入
        temp += cardPoints(left)
      } else {
        // 右边入，左边出
        temp += cardPoints(right)
        temp -= cardPoints(left)
        left -= 1
      }

      if (!flag && left == k - 1) {
        flag = true
      }

      if (flag) {
        // 更新
        ans = Math.max(temp, ans)
        right -= 1
      } else {
        left += 1
      }
    }
    ans
  }

  /**
   * 逆向思维，求前后 k 长度的最大值，就是求 n-k 连续子数组的最小值
   * @param cardPoints
   * @param k
   * @return
   */
  def reverseMaxScore(cardPoints: Array[Int], k: Int): Int = {
    val m = cardPoints.length - k
    var s = 0
    for (i <- 0 until m) {
      s += cardPoints(i)
    }
    var total = s
    var min = s
    for (i <- m until cardPoints.length) {
      total += cardPoints(i)
      s += cardPoints(i) - cardPoints(i - m)
      min = Math.min(s, min)
    }

    total - min
  }



  def main(args: Array[String]): Unit = {
    println(maxScore(Array(1,2,3,4,5,6,1), 3) == 12)
    println(maxScore(Array(2,2,2), 2) == 4)
    println(maxScore(Array(9,7,7,9,8,7,9), 4) == 33)
    println(maxScore(Array(1,1000,1), 2) == 1001)
    println(maxScore(Array(1,79,80,1,1,1,200,1), 3) == 202)
  }
}

File: src/main/java/com/oycm/dp/e/best_divide/Solution_8.java
package com.oycm.dp.e.best_divide;

public class Solution_8 {

    /**
     * 1043. <a href="https://leetcode.cn/problems/partition-array-for-maximum-sum/description/">分隔数组以得到最大和</a> 1916
     *
     * @param arr
     * @param k
     * @return
     */
    public int maxSumAfterPartitioning(int[] arr, int k) {
        /*
        给你一个整数数组 arr，请你将该数组分隔为长度 最多 为 k 的一些（连续）子数组。
        分隔完成后，每个子数组的中的所有值都会变为该子数组中的最大值。
        返回将数组分隔变换后能够得到的元素最大和。本题所用到的测试用例会确保答案是一个 32 位整数。
         */
        int n = arr.length;
        int[] f = new int[n + 1];
        for (int i = 0; i < n; i++) {
            // [j, i] 最多 k 个元素 i - j + 1 <= k => i - j < k
            for (int j = i, mx = 0; j >= 0 && i - j < k; j--) {
                mx = Math.max(mx, arr[j]);
                f[i + 1] = Math.max(f[i + 1], f[j] + (i - j + 1) * mx);
            }
        }
        return f[n];
    }

    public int dfs(int i, int k, int[] arr, int[] memo) {
        if (i == 0) {
            return 0;
        }
        if (memo[i] != -1) {
            return memo[i];
        }
        int res = 0, mx = 0;
        // 最长为 k，分为一组
        for (int j = i - 1; j >= Math.max(0, i - k); j--) {
            mx = Math.max(mx, arr[j]);
            res = Math.max(res, dfs(j, k, arr, memo) + (i - j) * mx);
        }
        return memo[i] = res;

    }
}
